### PR TITLE
Document that Sys.rename works on directories too

### PR DESCRIPTION
**Pull request by jmid**
This PR documents existing `Sys.rename` functionality over directories:
```ocaml
# Sys.mkdir "/tmp/foo" 0x755;;
- : unit = ()
# Sys.rename "/tmp/foo" "/tmp/bar";;
- : unit = ()
# Sys.is_directory "/tmp/bar";;
- : bool = true
# Sys.is_directory "/tmp/foo";;
Exception: Sys_error "/tmp/foo: No such file or directory".
```

and over symlinks:
```ocaml
$ ln -s /tmp/bar /tmp/link
$ ocaml
# Sys.rename "/tmp/link" "/tmp/link2";;
- : unit = ()
# Sys.file_exists "/tmp/link";;
- : bool = false
# Sys.file_exists "/tmp/link2";;
- : bool = true
```

Digging a bit, I found, e.g., https://github.com/ocaml/ocaml/pull/1306 to make `Sys.rename` more POSIX compatible.
On that note, https://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html specifies `rename` to work over directories and symlinks too.

This documentation PR is intended as a first step.
I'll open a separate issue with some corner case differences I found for Windows.

### CODE REVIEW

**Comment by avsm** [comment]
I think this would really benefit from some tests in the test suite to exercise these corner case renames.  They would then run on all the operating systems in the CI precheck.

**Comment by jmid** [comment]
Good point about tests to confirm this behaviour across platforms :+1: 
In 4018635 I've added a positive and negative test of `Sys.rename`ing a directory.
I've created ocaml/ocaml#12073 to track Windows differences on two other corner cases.

I'm unsure how to create a sym (or hard) link for exercising `Sys.rename` on them
   (`Sys.command` + `Sys.os_type` dispatching? using `Unix` for a `Sys` test? ...)

**Comment by shindere** [comment]
Jan Midtgaard (2023/03/03 12:11 -0800):
> Good point about tests to confirm this behaviour across platforms :+1:
> In 4018635 I've added a positive and negative test of `Sys.rename`ing a directory.
> I've created ocaml/ocaml#12073 to track Windows differences on two other corner cases.
>
> I'm unsure how to create a sym (or hard) link for exercising `Sys.rename` on them
>    (`Sys.command` + `Sys.os_type` dispatching? using `Unix` for a `Sys` test? ...)

For the little I know about these matters, I'd definitely notmake a Sys
test depend on Unix because, in theory, it isallowed to have Sys while
Unix is disabled.

Isn'tthere a risk that the test has to also depend on the Windows
version?

**Comment by jmid** [comment]
Oh wow. It turns out renaming a directory fails on macOS and OpenBSD (the one I had immediate access to) while succeeding on Linux and Windows...

**Comment by shindere** [comment]
Jan Midtgaard (2023/03/03 13:39 -0800):
> Oh wow. It turns out renaming a directory fails on macOS and OpenBSD
> (the one I had immediate access to) while succeeding on Linux and
> Windows...

Which error messages / return codes do you get?

**Comment by jmid** [comment]
> Which error messages / return codes do you get?

AFAICS, OpenBSD and macOS both fail the added `Rename directory to a nonexisting directory` test with `Sys_error exception: Permission denied`. It passes on Linux, Windows, and FreeBSD though.

I mistakingly thought this was just a omitted by mistake, but now start to suspect that the "under-specification" was intentional. 

`Sys` does not offer a facility to rename a directory AFAICS, so I think it makes sense to have `Sys.rename` support it.

**Comment by xavierleroy** [comment]
Your new test at https://github.com/ocaml/ocaml/commit/4018635064b9ce897a1b0778bd5ab69eea47ecc5 is flaky: the directory is created with permissions `0x755` while you meant `0o755`.  I think the funky permissions cause the error on macOS.  With `0o755` everything goes well.

**Comment by jmid** [comment]
> Your new test at 4018635 is flaky: the directory is created with permissions `0x755` while you meant `0o755`.

Indeed, thanks! :sweat_smile: 

I'm still unsure how to test renaming of links in a platform agnostic manner.
22d838e thus proposes to specify just the renaming of directories that we now have tests for.

**Comment by jmid** [comment]
I've just pushed a cleaned up git history.

File: testsuite/tests/lib-sys/rename.reference [hunk 1]
```diff
@@ -2,3 +2,5 @@ Rename to nonexisting file: passed
 Rename to existing file: passed
 Renaming a nonexisting file: fails as expected
 Renaming to a nonexisting directory: fails as expected
+Rename directory to a nonexisting directory: passed
+Rename a nonexisting directory: fails as expected
```
Comment (avsm) [inline testsuite/tests/lib-sys/rename.reference:6]: I think it would be worth testing renaming to an existing directory as well. There are some really weird corner cases (especially involving hardlinks) with filesystems, but just the simple case of ensuring that an existing target directory that contains a file with normal permissions is removed would be worthwhile.

Comment (jmid) [reply testsuite/tests/lib-sys/rename.reference:6]: OK :+1:  5370e4a adds a negative test to rename to an existing, non-empty target directory.

**Comment by jmid** [comment]
In the branch https://github.com/jmid/ocaml/tree/sys-rename-win-fixes I've now added
- the two corner-case tests currently failing on Windows reported in #12073 
- a proposed fix to `caml_win32_rename`

It is built on top of this PR. I can either create a separate PR with those or merge them into this one if you prefer.

**Review by MisterDA — APPROVED** [review]
The new tests are good, thanks. I think this can be merged as-is.

> moving it between (parent) directories if needed

Do we need to test this behavior? I can't think of a case where that wouldn't work if the user has permissions on the destination.

**Review by gasche — APPROVED** [review]
Approved on @MisterDA's behalf.
